### PR TITLE
KEP-4330:  moving apiextensions-apiserver featuregates to versioned.

### PR DESCRIPTION
**Pull request by liangyuanpeng**
#### What type of PR is this?

/kind cleanup
#### What this PR does / why we need it:
Part of https://github.com/kubernetes/kubernetes/issues/127159
Part of https://github.com/kubernetes/kubernetes/issues/126926

/assign @Jefftree

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by liangyuanpeng** [comment]
Seems like it's the problem of content order

```
$ hack/verify-featuregates.sh
found 39 features in FeatureSpecMap var defaultKubernetesFeatureGates in file: /home/runner/work/lanactions/lanactions/kubernetes/pkg/features/kube_features.go
found 31 features in FeatureSpecMap var defaultKubernetesFeatureGates in file: /home/runner/work/lanactions/lanactions/kubernetes/staging/src/k8s.io/apiserver/pkg/features/kube_features.go
found 3 features in FeatureSpecMap of func featureGates in file: /home/runner/work/lanactions/lanactions/kubernetes/staging/src/k8s.io/component-base/logs/api/v1/kube_features.go
found 1 features in FeatureSpecMap of func featureGates in file: /home/runner/work/lanactions/lanactions/kubernetes/staging/src/k8s.io/component-base/metrics/features/kube_features.go
found 1 features in FeatureSpecMap var cloudPublicFeatureGates in file: /home/runner/work/lanactions/lanactions/kubernetes/staging/src/k8s.io/controller-manager/pkg/features/kube_features.go
found 110 features in FeatureSpecMap var defaultVersionedKubernetesFeatureGates in file: /home/runner/work/lanactions/lanactions/kubernetes/pkg/features/versioned_kube_features.go
found 2 features in FeatureSpecMap var defaultVersionedKubernetesFeatureGates in file: /home/runner/work/lanactions/lanactions/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go
found 1 features in FeatureSpecMap var defaultVersionedKubernetesFeatureGates in file: /home/runner/work/lanactions/lanactions/kubernetes/staging/src/k8s.io/apiserver/pkg/features/kube_features.go
panic: detected diff in versioned feature list (test/featuregates_linter/test_data/versioned_feature_list.yaml), diff: 
          []cmd.featureInfo{
                ... // 11 identical elements
                {Name: "CPUManagerPolicyBetaOptions", VersionedSpecs: {{Default: true, PreRelease: "Beta", Version: "1.23"}}},
                {Name: "CPUManagerPolicyOptions", VersionedSpecs: {{PreRelease: "Alpha", Version: "1.22"}, {Default: true, PreRelease: "Beta", Version: "1.23"}}},
        -       {
        -               Name: "CRDValidationRatcheting",
        -               VersionedSpecs: []cmd.featureSpec{
        -                       {PreRelease: "Alpha", Version: "1.28"},
        -                       {Default: true, PreRelease: "Beta", Version: "1.30"},
        -               },
        -       },
                {Name: "CronJobsScheduledAnnotation", VersionedSpecs: {{Default: true, PreRelease: "Beta", Version: "1.28"}}},
                {Name: "CrossNamespaceVolumeDataSource", VersionedSpecs: {{PreRelease: "Alpha", Version: "1.26"}}},
                {Name: "CSIMigrationPortworx", VersionedSpecs: {{PreRelease: "Alpha", Version: "1.23"}, {PreRelease: "Beta", Version: "1.25"}, {Default: true, PreRelease: "Beta", Version: "1.31"}}},
                {Name: "CSIVolumeHealth", VersionedSpecs: {{PreRelease: "Alpha", Version: "1.21"}}},
        -       {
        -               Name: "CustomResourceFieldSelectors",
        -               VersionedSpecs: []cmd.featureSpec{
        -                       {PreRelease: "Alpha", Version: "1.30"},
        -                       {Default: true, PreRelease: "Beta", Version: "1.31"},
        -               },
        -       },
                {Name: "DevicePluginCDIDevices", VersionedSpecs: {{PreRelease: "Alpha", Version: "1.28"}, {Default: true, PreRelease: "Beta", Version: "1.29"}, {Default: true, LockToDefault: true, PreRelease: "GA", Version: "1.31"}}},
                {Name: "DisableAllocatorDualWrite", VersionedSpecs: {{PreRelease: "Alpha", Version: "1.31"}}},
                ... // 87 identical elements
                {Name: "WinDSR", VersionedSpecs: {{PreRelease: "Alpha", Version: "1.14"}}},
                {Name: "WinOverlay", VersionedSpecs: {{PreRelease: "Alpha", Version: "1.14"}, {Default: true, PreRelease: "Beta", Version: "1.20"}}},
        +       {
        +               Name: "CRDValidationRatcheting",
        +               VersionedSpecs: []cmd.featureSpec{
        +                       {PreRelease: "Alpha", Version: "1.28"},
        +                       {Default: true, PreRelease: "Beta", Version: "1.30"},
        +               },
        +       },
        +       {
        +               Name: "CustomResourceFieldSelectors",
        +               VersionedSpecs: []cmd.featureSpec{
        +                       {PreRelease: "Alpha", Version: "1.30"},
        +                       {Default: true, PreRelease: "Beta", Version: "1.31"},
        +               },
        +       },
          }


goroutine 1 [running]:
k8s.io/kubernetes/test/featuregates_linter/cmd.verifyFeatureListFunc(0xc0001b6200?, {0x695a38?, 0x4?, 0x6959e8?})
        /home/runner/work/lanactions/lanactions/kubernetes/test/featuregates_linter/cmd/feature_gates.go:102 +0x85
github.com/spf13/cobra.(*Command).execute(0xc0001ce308, {0x8c2e80, 0x0, 0x0})
        /home/runner/work/lanactions/lanactions/kubernetes/vendor/github.com/spf13/cobra/command.go:989 +0xa91
github.com/spf13/cobra.(*Command).ExecuteC(0x89ed40)
        /home/runner/work/lanactions/lanactions/kubernetes/vendor/github.com/spf13/cobra/command.go:1117 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
        /home/runner/work/lanactions/lanactions/kubernetes/vendor/github.com/spf13/cobra/command.go:1041
k8s.io/kubernetes/test/featuregates_linter/cmd.Execute()
        /home/runner/work/lanactions/lanactions/kubernetes/test/featuregates_linter/cmd/root.go:32 +0x1a
main.main()
        /home/runner/work/lanactions/lanactions/kubernetes/test/featuregates_linter/main.go:22 +0xf
exit status 2
```

**Comment by liangyuanpeng** [comment]
Got it, It needs to be sorted manually in versioned_feature_list.yaml, i'll update it later.

**Comment by Jefftree** [comment]
/ok-to-test

> Got it, It needs to be sorted manually in `versioned_feature_list.yaml`, i'll update it later.

Quick question, did you end up having to manually edit the versioned_feature_list.yaml file? If so that's a bug in our verification machinery we'll need to look into.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go [hunk 1]
```diff
@@ -45,12 +48,26 @@ const (
 
 func init() {
 	utilfeature.DefaultMutableFeatureGate.Add(defaultKubernetesFeatureGates)
+	utilfeature.DefaultMutableFeatureGate.AddVersioned(defaultVersionedKubernetesFeatureGates)
```
Comment (Jefftree) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go:51]: `runtime.Must(utilfeature.DefaultMutableFeatureGate...`

**Comment by liangyuanpeng** [comment]
> can we resolve the issues we're seeing with versioned gates, compatibility mode, and testing (like https://github.com/kubernetes/kubernetes/issues/127156 and https://github.com/kubernetes/kubernetes/issues/127174) before moving more gates into this system?

/hold 
for https://github.com/kubernetes/kubernetes/pull/127200#issuecomment-2334180949


> You should be able to update that list via hack/update-featuregates.sh

Ohh, I missed it,  i think we can add some NOTE in `test/featuregates_linter/test_data/versioned_feature_list.yaml` to tell people that it's updated automatically and not edited manually, like `Do not edit it, this is generate by hack/update-featuregates.sh`, WDYT?

**Comment by Jefftree** [comment]
With https://github.com/kubernetes/kubernetes/issues/127174 fixed,  we should be able to continue work on this. Might need a rebase since feature gates have changed.

**Comment by Jefftree** [comment]
/close
Features have already been ported.

**Pull request closed without merging**